pkg/autoscaler/autoscaler.go

```diff
@@ -3,126 +3,208 @@ package autoscaler
-	threshold      int64
+	logger                  logger.Logger
+	namespace               string
+	resourceScaler          scaler_types.ResourceScaler
+	scaleInterval           time.Duration
+	inScaleToZeroProcessMap map[string]bool
+	groupKind               string
+	customMetricsClientSet  custommetricsv1.CustomMetricsClient
 }
 
 func NewAutoScaler(parentLogger logger.Logger,
 	resourceScaler scaler_types.ResourceScaler,
+	customMetricsClientSet custommetricsv1.CustomMetricsClient,
 	options scaler_types.AutoScalerOptions) (*Autoscaler, error) {
-	childLogger := parentLogger.GetChild("autoscale")
+	childLogger := parentLogger.GetChild("autoscaler")
 	childLogger.DebugWith("Creating Autoscaler",
-		"Namespace", options.Namespace,
-		"MetricName", options.MetricName)
+		"options", options)
 
 	return &Autoscaler{
-		logger:         childLogger,
-		namespace:      options.Namespace,
-		metricsMap:     make(resourceMetricTypeMap),
-		resourceScaler: resourceScaler,
-		metricName:     options.MetricName,
-		windowSize:     options.ScaleWindow,
-		scaleInterval:  options.ScaleInterval,
-		threshold:      options.Threshold,
-		metricsChannel: make(chan metricEntry, 1024),
+		logger:                  childLogger,
+		namespace:               options.Namespace,
+		resourceScaler:          resourceScaler,
+		scaleInterval:           options.ScaleInterval,
+		groupKind:               options.GroupKind,
+		customMetricsClientSet:  customMetricsClientSet,
+		inScaleToZeroProcessMap: make(map[string]bool),
 	}, nil
 }
 
-func (as *Autoscaler) checkResourceToScale(t time.Time, activeResources []scaler_types.Resource) {
-	for _, resourceName := range activeResources {
+func (as *Autoscaler) getMetricNames(resources []scaler_types.Resource) []string {
+	var metricNames []string
+	for _, resource := range resources {
+		for _, scaleResource := range resource.ScaleResources {
+			metricNames = append(metricNames, scaleResource.GetKubernetesMetricName())
+		}
+	}
+	metricNames = common.UniquifyStringList(metricNames)
+	as.logger.DebugWith("Got metric names", "metricNames", metricNames)
+	return metricNames
+}
 
-		// currently only one type of metric supported from a platform configuration
-		resourceMetrics := as.metricsMap[resourceName][as.metricName]
+func (as *Autoscaler) getResourcesMetrics(metricNames []string) (map[string]map[string]int, error) {
+	resourcesMetricsMap := make(map[string]map[string]int)
 
-		// this will give out the greatest delta
-		var minMetric *metricEntry
-		for idx, metric := range resourceMetrics {
+	schemaGroupKind := schema.GroupKind{Group: "", Kind: as.groupKind}
+	resourceLabels := labels.Everything()
+	c := as.customMetricsClientSet.NamespacedMetrics(as.namespace)
 
-			if metric.value <= as.threshold && minMetric == nil {
-				minMetric = &resourceMetrics[idx]
-			} else if metric.value > as.threshold {
-				minMetric = nil
+	for _, metricName := range metricNames {
+		cm, err := c.GetForObjects(schemaGroupKind,
```

Comment:
Can you add some comments here - what does GetForObjects return and why?
Would be great to put lots of comments here to tell the story